src/platform_impl/macos/progress_bar.rs

```diff
@@ -0,0 +1,156 @@
+use std::sync::Once;
+
+use cocoa::{
+  base::{id, nil},
+  foundation::{NSArray, NSPoint, NSRect, NSSize},
+};
+use objc::{
+  declare::ClassDecl,
+  runtime::{Class, Object, Sel, NO},
+};
+
+use crate::window::{ProgressBarState, ProgressState};
+
+/// Set progress indicator in the Dock.
+pub fn set_progress_indicator(progress_state: ProgressBarState) {
```

Comment:
the implementation actually looks similar. Electron is also using a custom redraw function: https://github.com/electron/electron/blob/b142fce229c5821083ad6b7c9aa6aa5790032127/shell/browser/native_window_mac.mm#L60

I've tried without the custom draw function and it looks a little weird, it only refreshes on the second update.

Reply:
Though I can see the native one looks better.

Reply:
custom:
<img width="80" alt="Screenshot 2023-07-12 at 13 30 59" src="https://github.com/tauri-apps/tao/assets/20051258/a693ab10-958e-461b-a4aa-400e0032b860">
native:
<img width="80" alt="Screenshot 2023-07-12 at 13 31 34" src="https://github.com/tauri-apps/tao/assets/20051258/63697106-f074-4259-94d5-12b9969017cf">

Reply:
@pewsheen the only issue is I can't seem to synchronize it. To use you just need to swap your custom class with NSProgressIndicator and call `setIndeterminate: NO`, but increasing the progress is never done immediately, I need to update to a higher value again to trigger it. Decreasing the progress works fine though.

Reply:
Also indeterminate mode doesn't seem to make a change (with both classes).

Reply:
Yeah I tried the animation thing too. I didn't check the threaded animation thing though, weird :D

Reply:
Indeed it works, nice :) maybe the issue is we're updating the value on the main thread?

Reply:
> the implementation actually looks similar. Electron is also using a custom redraw function: [electron/electron@`b142fce`/shell/browser/native_window_mac.mm#L60](https://github.com/electron/electron/blob/b142fce229c5821083ad6b7c9aa6aa5790032127/shell/browser/native_window_mac.mm#L60)
> 
> I've tried without the custom draw function and it looks a little weird, it only refreshes on the second update.

my bad, didn't notice but if we are drawing it manually, I believe we can support the different states like windows instead of treating all variants as Normal?

Reply:
> Indeed it works, nice :) maybe the issue is we're updating the value on the main thread?

I tried using `DispatchQueue` to send it to other threads and it still not drawing until the next update. 

> my bad, didn't notice but if we are drawing it manually, I believe we can support the different states like windows instead of treating all variants as Normal?

Yes, we can easily change styles like colors for different states. But the animation would be another story 👀

Reply:
I created a SwiftUI project from Xcode and tested the native `NSProgressIndicator` on the dock tile, the behavior is the same.

It feels like they don't encourage us to apply animation on the dock tile.  I saw some implementations that animated the dock icon by calling `display()` in each animation frame.

Reply:
> Yes, we can easily change styles like colors for different states. But the animation would be another story 👀

Yeah I think modifying color is enough

Reply:
> > Yes, we can easily change styles like colors for different states. But the animation would be another story eyes
> 
> Yeah I think modifying color is enough

I agree.

Reply:
Ok! I pushed a commit that can change colors for `paused` and `error` states.

I'm really bad at styling, so feel free to edit it :joy: